### PR TITLE
[redux-define] add types

### PR DESCRIPTION
**Pull request by ensconced**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] [Add or edit tests](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#my-package-teststs) to reflect the change.
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] [Run `npm test <package to test>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#running-tests).

Select one of these and delete the others:

If adding a new definition:
- [x] The package does not already provide its own types, or cannot have its `.d.ts` files generated via `--declaration`
- [x] If this is for an npm package, match the name. If not, do not conflict with the name of an npm package.
- [x] Create it with `dts-gen --dt`, not by basing it on an existing project.
- [x] Represents shape of module/library [correctly](https://www.typescriptlang.org/docs/handbook/declaration-files/library-structures.html)
- [x] `tslint.json` [should contain](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#linter-tslintjson) `{ "extends": "@definitelytyped/dtslint/dt.json" }`, and no additional rules.
- [x] `tsconfig.json` [should have](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#tsconfigjson) `noImplicitAny`, `noImplicitThis`, `strictNullChecks`, and `strictFunctionTypes` set to `true`.

### CODE REVIEW

**Comment by peterblazejewicz** [comment]
This looks cool. The only nitpick is the problem with dev-exp enforcing `as const` when coding. Can you look around to figure out how to force TS to infer values for string templates without passing that burden on the user? Ask on discord TS channel, best place

**Comment by ensconced** [comment]
@peterblazejewicz Thanks for reviewing! In the current version, you should be able to pass arrays without `as const` without any error, but the types you get back for the actions will be wider e.g. `CREATE_TODO_${string}`. As far as I know, it won't be possible to get exact string types for the actions without using `as const` (when subactions are being used). But I will check up on that.

**Comment by ensconced** [comment]
turns out it is possible, fixed :)

File: types/redux-define/index.d.ts [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+    Namespace extends string | undefined = undefined,
+> = SubActions extends string[]
+    ? PlainAction<OwnAction, Namespace> & SubActionProps<SubAction, SubActions, Namespace, OwnAction>
+    : PlainAction<OwnAction, Namespace>;
+
+export function defineAction<OwnAction extends string>(actionType: OwnAction): Action<OwnAction>;
+export function defineAction<OwnAction extends string, SubAction extends string, S extends SubAction[]>(
+    actionType: OwnAction,
+    subactions: S,
+): Action<OwnAction, SubAction, S>;
```
Comment (peterblazejewicz) [inline types/redux-define/index.d.ts:45]:
```suggestion
export function defineAction<OwnAction extends string, SubAction extends string, SubActions extends SubAction[]>(
    actionType: OwnAction,
    subactions: SubActions,
): Action<OwnAction, SubAction, SubActions>;
```
should it be `SubActions` for consistency?

File: types/redux-define/index.d.ts [hunk 2]
```diff
@@ -0,0 +1,82 @@
+// Type definitions for redux-define 1.1
+// Project: https://github.com/smeijer/redux-define
+// Definitions by: Joe Barnett <https://github.com/ensconced>
+// Definitions: https://github.com/DefinitelyTyped/DefinitelyTyped
+// Minimum TypeScript Version: 4.1
+
+export as namespace ReduxDefine;
```
Comment (amcasey) [inline types/redux-define/index.d.ts:7]: Why export as namespace?

File: types/redux-define/index.d.ts [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+    actionType: OwnAction,
+    subactions: SubActions,
+    namespace: Namespace,
+): Action<OwnAction, SubAction, SubActions, NamespaceString<Namespace>>;
+export function defineAction<OwnAction extends string, Namespace extends string | Action>(
+    actionType: OwnAction,
+    namespace: Namespace,
+): Action<OwnAction, string, undefined, NamespaceString<Namespace>>;
+
+export function defineActionMethod<Parent extends Action, OwnAction extends string>(
```
Comment (amcasey) [inline types/redux-define/index.d.ts:61]: The underlying module doesn't appear to export a function called `defineActionMethod`.  If this is just a helper, consider dropping the `export`.

**Review by amcasey — CHANGES_REQUESTED** [review]

File: types/redux-define/index.d.ts (hunk 2 above)
Comment (peterblazejewicz) [reply types/redux-define/index.d.ts:7]: It has UMD distribution, can be used as global script, sample:
https://codepen.io/blazejewicz/pen/GRQMpoo

File: types/redux-define/index.d.ts (hunk 3 above)
Comment (ensconced) [reply types/redux-define/index.d.ts]: The tests didn't allow removing the `export` directly, that gives this error:

```
ll declarations in this module are exported automatically. Prefer to explicitly write 'export' for clarity.
```

But I have rewritten this using arrow function types which are intersected to give the overload.

**Review by peterblazejewicz — APPROVED** [review]
LGTM!
@ensconced thanks!

**Pull request merged**